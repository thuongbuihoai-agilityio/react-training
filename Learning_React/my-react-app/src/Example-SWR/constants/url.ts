const BASE_URL = "http://localhost:1234";
const TASK_URL = `${BASE_URL}/task`;

export { BASE_URL, TASK_URL };
