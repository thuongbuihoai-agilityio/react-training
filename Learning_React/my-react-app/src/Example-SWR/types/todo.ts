export interface TodoType {
  id: string;
  title: string;
}

export interface TodoProps {
  todo: TodoType;
  deleteTodo: (id: string) => void;
}

export interface ModalUpdateProps {
  todo: TodoType;
  hideModalUpdate: Function;
  isReload: Boolean;
  setIsReLoad: Function;
}