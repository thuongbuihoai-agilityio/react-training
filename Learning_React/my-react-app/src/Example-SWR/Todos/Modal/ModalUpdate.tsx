import React, { useState } from "react";
import Button from "../../components/Button/Button";
import useTodo from "../../hooks/useTodo";
import { ModalUpdateProps, TodoType } from "../../types/todo";

const ModalUpdate: React.FC<ModalUpdateProps> = ({todo, isReload, setIsReLoad, hideModalUpdate}) => {
  const {updateTodo}  = useTodo();
  const [todoEdit, setTodoEdit] = useState(todo);

  const update = (id: string, data: TodoType) => {
    updateTodo(id, data);
    setIsReLoad(!isReload);
    hideModalUpdate(false);
  }

  const handleChange = (event: { target: { value: {}; name: string; } }) => {
    const value = event.target.value;
    const key = event.target.name;
    setTodoEdit({...todoEdit, [key]: value});
  }

  return (
    <div className="modals" id="bookDeleteModal">
      <div className="modal-dialog">
        <div className="modal-content">
          <div className="modal-header">
            <h4 className="modal-title" id="productModalLabel">Edit Task</h4>
          </div>
          <div className="modal-body">
            <div className="form-control">
              <label htmlFor="">Title:</label>
              <input data-testid="update-input" className="form__input" type="text" onChange={handleChange} value={todoEdit.title} name="title" />
            </div>
          </div>
          <div className="modal-footer">
            <Button value="Submit" className="btn btn__submit" onClick={() => update(todoEdit.id, todoEdit)} />
            <Button value="Cancel" className="btn btn__cancel" onClick={() => hideModalUpdate(false)} />
          </div>
        </div>
      </div>
    </div>
  );
}

export default ModalUpdate;
