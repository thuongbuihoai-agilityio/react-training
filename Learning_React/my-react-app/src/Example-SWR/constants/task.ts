import { TodoType } from "../types/todo";

const TASK_MOCKING: TodoType = {
  id: "1",
  title: "Example create Todo App (20220522)"
};

export {TASK_MOCKING};