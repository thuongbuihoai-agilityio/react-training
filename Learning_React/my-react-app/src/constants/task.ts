import { TodoType } from "../Example-SWR/types/todo";

const TASK_MOCKING: TodoType = {
  id: "1",
  title: "Example create Todo App (20220522)"
};

export {TASK_MOCKING};