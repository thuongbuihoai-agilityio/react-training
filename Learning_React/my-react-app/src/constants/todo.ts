import { TodoProps } from "../Example-SWR/types/todo";

const TODO_CONSTANTS: TodoProps = {
  todo: {
    id: "a56403e9-8163-451c-a17b-6bd2fa3f389a",
    title: "Example create Todo App (20220522)",
  },
  deleteTodo: () => {}
};

export {
  TODO_CONSTANTS
}