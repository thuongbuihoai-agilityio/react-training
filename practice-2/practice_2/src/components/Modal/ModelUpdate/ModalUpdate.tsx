import axios from "axios";
import React, { useContext, useState } from "react";
import { useLocation } from "react-router-dom";
import useCategory from "../../../hooks/useCategory";
import { CategoriesProps } from "../../../types/categories";
import { ProductUpdateProps, ProductProps } from "../../../types/product";
import { SUCCESS_MSG } from "../../../constants/message";
import { ProductListContext } from "../../../App";
import { ModalUpdateProps } from "types/modal";
import { PRODUCT_CRUD } from "../../../constants/url";
import "./modalUpdate.css";

const ModalUpdate: React.FC<ModalUpdateProps> = ({ id, hideModalUpdate, onChangeProductDetail }) => {
  const location = useLocation();
  const product: ProductProps = (location.state as ProductUpdateProps).product;
  const [productEdit, setProductEdit] = useState<ProductProps>(product);

  const categories = useCategory();
  const [selectedFile, setSelectedFile] = useState([]);

  const handleUpdateProduct = async (id: string) => {
    for (let i = 0; i < selectedFile.length; i++) {
      productEdit.images.push("src/assets/images/" + selectedFile[i].name);
    }
      await axios ({
        method: "put",
        url: PRODUCT_CRUD + id,
        data:{...productEdit},
      })
      .then(function (response) {
        onChangeProductDetail(response.data);
        alert(SUCCESS_MSG.MESSAGE_UPDATE_PRODUCT);
        hideModalUpdate(false);
      })
      .catch(function (error) {
        alert(error);
      });
    }

  const handleChange = (event: { target: { value: {}; name: string; } }) => {
    const value = event.target.value;
    const key = event.target.name;
    setProductEdit({...productEdit, [key]: value});
  }

  const imageChange = (event: React.ChangeEvent) => {
    const target= event.target;
    if(target.files) {
      for (let i = 0; i < event.target.files.length; i++) {
        setSelectedFile(selectedFile => [...selectedFile, target.files[i]] as any);
      }
    }
  }

  return (
    <div className="modals" id="bookDeleteModal">
      <div className="modal-dialog">
        <div className="modal-content">
          <div className="modal-header">
            <h4 className="modal-title" id="productModalLabel">Edit Product Information</h4>
          </div>
          <div className="modal-body">
            <div className="form-control">
              <label htmlFor="">Product name: </label>
              <input className="form__input" type="text" onChange={handleChange} name="name" value={productEdit.name} />
            </div>
            <div className="form-control">
              <label htmlFor="">Description: </label>
              <textarea className="form__text" value={productEdit.description} onChange={handleChange} name="description" id="" cols={30} rows={5}></textarea>
            </div>
            <div className="form-control">
              <label htmlFor="">Categories: </label>
              <select className="form__select" name="categoryId" onChange={handleChange} id="">
                {categories.map((cate: CategoriesProps, index: number) =>
                  <option key={index} value={cate.id} selected={productEdit.categoryId == cate.id}>{cate.name}</option>
                )}
              </select>
            </div>
            <div id="form__number" className="form-control">
              <div className="form__price">
                <label htmlFor="">Price: </label>
                <input className="modal__input" type="number" min="0" name="price" onChange={handleChange} value={productEdit.price} />
              </div>
              <div className="form__quantity">
                <label htmlFor="">Quantity: </label>
                <input className="modal__input" type="number" min="0" name="quantity" onChange={handleChange} value={productEdit.quantity} />
              </div>
            </div>
            <div className="form-control">
              <div className="form__img--list">
                {product.images.map((img: string, index: number) =>
                  <img key={index} className="form__img" src={img} />
                )}
                <input type="file" id="file" multiple name="images" onChange={imageChange} className="form__input--img"/>
              </div>
            </div>
          </div>
          <div className="modal-footer">
            <button className="btn btn__submit" onClick={() => handleUpdateProduct(id)}>Submit</button>
            <button className="btn btn__cancel" onClick={() => hideModalUpdate(false)}>Cancel</button>
          </div>
        </div>
      </div>
    </div>
  );
}

export default ModalUpdate;
