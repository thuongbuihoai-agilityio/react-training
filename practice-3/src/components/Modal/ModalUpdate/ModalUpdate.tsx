import useSWR, { Key } from "swr";
import React, { useState } from "react";
import { ModalUpdateProps } from "@/types/modal";
import { CategoryProps } from "@/types/category";
import { Product } from "@/types/product";
import getBase64 from "@/helpers/getBase64";
import { get, update } from "@/helpers/fetchApi";
import { CATEGORIES_URL, PRODUCT_CRUD } from "@/constants/url";
import { SUCCESS_MSG } from "@/constants/message";
import { toast } from "react-toastify";
import "./modalUpdate.css";

const ModalUpdate: React.FC<ModalUpdateProps> = ({
  product,
  isReload,
  setIsReLoad,
  hideModalUpdate,
}) => {
  const key: Key = CATEGORIES_URL;
  const fetcher = () => get<Product[]>(CATEGORIES_URL);
  const { data, mutate } = useSWR(key, fetcher);
  const [selectedFile, setSelectedFile] = useState([]);
  const [productEdit, setProductEdit] = useState(product);
  const updateProduct = async (id: string, productData: Product) => {
    const productEdit: Product = {
      id: productData.id,
      categoryId: productData.categoryId,
      name: productData.name,
      price: +productData.price,
      quantity: +productData.quantity,
      description: productData.description,
      images: productData.images,
    };
    update(`${PRODUCT_CRUD}/${id}`, productEdit);
    mutate();
    setIsReLoad(!isReload);
    toast.success(SUCCESS_MSG.MESSAGE_UPDATE_PRODUCT);
  };

  const handleUpdateProduct = async (id: string, productEdit: Product) => {
    for (let i = 0; i < selectedFile.length; i++) {
      productEdit.images.push(selectedFile[i]);
    }
    updateProduct(id, productEdit);
    hideModalUpdate();
  };

  const handleChange = (event: { target: { value: {}; name: string } }) => {
    const value = event.target.value;
    const key = event.target.name;
    setProductEdit({ ...productEdit, [key]: value });
  };

  const imageChange = async (event: { target: { files: any } }) => {
    const files = event.target.files;
    if (files) {
      for (let i = 0; i < files.length; i++) {
        const imageSrc = await getBase64(files[i]);
        setSelectedFile((selectedFile) => [...selectedFile, imageSrc] as any);
      }
    }
  };

  const handleDeleteImage = (event: { target: Element | any}) => {
    const target = event.target as Element;
    const indexOfArr = productEdit.images.findIndex(
      (item: string) => item == (target as HTMLInputElement).dataset.id
    );
    const indexOf = selectedFile.findIndex(
      (item: string) => item == (target as HTMLInputElement).dataset.id
    );
    selectedFile.splice(indexOf, 1);
    productEdit.images.splice(indexOfArr, 1);
    setSelectedFile([...selectedFile]);
  };

  return (
    <div data-testid="modal-update" className="modal-update">
      <div className="modal-dialog-modalUpdate">
        <div className="modal-content">
          <div className="modal-header">
            <h4 className="modal-heading" id="productModalLabel">
              Edit Product Information
            </h4>
          </div>
          <div className="modal-body">
            <div className="form-control">
              <label htmlFor="">Product name: </label>
              <input
                data-testid="change-value-name"
                className="form__input"
                type="text"
                name="name"
                value={productEdit?.name}
                onChange={handleChange}
              />
            </div>
            <div className="form-control">
              <label htmlFor="">Description: </label>
              <textarea
                className="form__text"
                cols={30}
                rows={5}
                value={productEdit?.description}
                name="description"
                onChange={handleChange}
              ></textarea>
            </div>
            <div className="form-control">
              <label htmlFor="">Categories: </label>
              <select
                className="form__select"
                name="categoryId"
                onChange={handleChange}
              >
                {data?.map(({id, name}: CategoryProps, index: number) => (
                  <option
                    key={index}
                    value={id}
                    selected={productEdit?.categoryId == id}
                  >
                    {name}
                  </option>
                ))}
              </select>
            </div>
            <div id="form__number" className="form-control">
              <div className="form-control">
                <label htmlFor="">Price: </label>
                <input
                  className="modal__input"
                  type="number"
                  min={0}
                  value={productEdit?.price}
                  name="price"
                  onChange={handleChange}
                />
              </div>
              <div className="form-control">
                <label htmlFor="">Quantity: </label>
                <input
                  className="modal__input"
                  type="number"
                  min={0}
                  value={productEdit?.quantity}
                  name="quantity"
                  onChange={handleChange}
                />
              </div>
            </div>
            <div className="form-control">
              <div className="form__img--list">
                {productEdit?.images?.map((img: string, index: number) => (
                  <img
                    data-testid="delete-image"
                    key={index}
                    className="form__img"
                    data-id={img}
                    src={img}
                    onClick={handleDeleteImage}
                  />
                ))}
                {selectedFile.length > 0 &&
                  selectedFile.map((src, key: number) => (
                    <img
                      data-testid="after-change-file"
                      key={key}
                      className="form__img"
                      data-id={src}
                      src={src}
                      onClick={handleDeleteImage}
                    />
                  ))}
                <input
                  data-testid="change-file"
                  className="form__input--img"
                  type="file"
                  id="file"
                  multiple
                  name="images"
                  onChange={imageChange}
                />
              </div>
            </div>
          </div>
          <div className="modal-footer-modalUpdate">
            <button
              data-testid="btn-no-modalUpdate"
              className="btn btn__no"
              onClick={hideModalUpdate}
            >
              Cancel
            </button>
            <button
              data-testid="btn-yes-modalUpdate"
              className="btn btn__yes"
              onClick={() => handleUpdateProduct(product.id, productEdit)}
            >
              Submit
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default ModalUpdate;
