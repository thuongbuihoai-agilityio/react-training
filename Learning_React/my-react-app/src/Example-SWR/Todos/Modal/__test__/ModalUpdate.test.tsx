import { act, fireEvent, render, screen } from "@testing-library/react";
import { TODO_CONTENT } from "../../../constants/todo";
import ModalUpdate from "../ModalUpdate";

describe("Button component", () => {
  test("Should render pop when clicking button", async () => {
    const isOpenModalUpdate = jest.fn();
    render (
      <ModalUpdate hideModalUpdate={isOpenModalUpdate}
        todo={TODO_CONTENT}
        isReload={true}
        setIsReLoad={() => {}} 
      />
    );

    // test for event click Submit and Cancel button on pop confirm
    const clickBtnSubmit = screen.getByRole("button", {name: /Submit/i});
    fireEvent.click(clickBtnSubmit);
    expect(isOpenModalUpdate).toHaveBeenCalled();

    const clickBtnCancel = screen.getByRole("button", {name: /Cancel/i});
    fireEvent.click(clickBtnCancel);
    expect(isOpenModalUpdate).toHaveBeenCalled();
  });

  test("allows user to update toto", async () => {
    const loadData = jest.fn();
    act(() => {
      render (
        <ModalUpdate hideModalUpdate={() => {}}
          todo={TODO_CONTENT}
          isReload={true}
          setIsReLoad={loadData}
        />
      );
      const clickBtnSubmit = screen.getByRole("button", {name: /Submit/i});
      const inputElement = screen.getByTestId("update-input") as HTMLInputElement;
      fireEvent.change(inputElement, {target: { value: "Writing unit test for update modal" } });
      fireEvent.click(clickBtnSubmit);
      expect(inputElement.value).toBe("Example create Todo App (20220522)");
    });
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
    <ModalUpdate hideModalUpdate={() => {}}
      todo={TODO_CONTENT}
      isReload={true}
      setIsReLoad={() => {}} 
    />);
    expect(asFragment()).toMatchSnapshot();
  });
});