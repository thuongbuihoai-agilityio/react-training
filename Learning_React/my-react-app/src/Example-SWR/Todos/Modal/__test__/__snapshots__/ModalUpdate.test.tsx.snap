// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="modals"
    id="bookDeleteModal"
  >
    <div
      class="modal-dialog"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h4
            class="modal-title"
            id="productModalLabel"
          >
            Edit Task
          </h4>
        </div>
        <div
          class="modal-body"
        >
          <div
            class="form-control"
          >
            <label
              for=""
            >
              Title:
            </label>
            <input
              class="form__input"
              data-testid="update-input"
              name="title"
              type="text"
              value="Example create Todo App (20220522)"
            />
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="btn btn__submit"
          >
            Submit
          </button>
          <button
            class="btn btn__cancel"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
